adrs/006-runtime-exit.md

```diff
@@ -0,0 +1,109 @@
+directory. Through this lens, it's appropriate for it to handle exits as well.
+
+This also offers a clear, consistent mechanism for overriding exit behavior -
+that is, overriding the Host. The `Cli` class already supports a custom
+`Host`, and the tests include a `MockConsoleHost` used for these purposes.
+It would be natural to extend `MockConsoleHost` to implement a test-only
+behavior for exits as well.
+
+### Why an EventEmitter?
+
+The event emitter interface is largely motivated by an interest in delegating
+exit behavior to the `Commander`.
+
+Given you *are* going to delegate to the `Commander`, the alternative to
+an event is injecting it as a dependency to the `Runtime`. Events allow
+the runtime to be unaware of the commander, at the cost of the commander
+being unable to "yield" data back to the runtime.
+
+Unfortunately, there are already reasons to inject the commander into the
+runtime. In particular, the commander handles prompting, because it handles
+the `readline` interface. This decision was made because readline requires
+asynchronous initialization, and because it's higher level than what the
+host provides. That decision isn't set in stone, but it *is* really convenient.
+
+While it hasn't been implemented yet, the runtime will need to request input
+from the commander *eventually* - so we might as well inject it now and avoid
+two interfaces.
+
+But we could also inject the host into the runtime, and have it call `Host#exit`
+directly. The alternative to this is implementing proxy methods on the
+commander whenever the runtime needs to access anything from the host. But
+the host contains a *lot* of functionality, and effectively adding all of the
+host's functionality to the commander muddies its interface.
+
+## Decision
+
+1. The `Host` will be injected into the `Runtime`, where its `exit` method will
+   be called directly. This will follow a pattern which should become more
+   common over time.
+2. The `Runtime` will *not* inherit from `EventEmitter`, instead preferring
+   to call methods on an injected `Commander` instance. This will create one
+   consistent way to call back to the `Commander` that supports "yielding".
+3. Runtime options will no longer accept an exit handler. Instead, that
+   behavior will be delegated to the already injected Host.
+4. `ConsoleHost#exit` will call `process.exit` for now.
+5. `MockConsoleHost#exit` will throw a `MockExit` error, maintaining the
+   current structure of the tests.
+6. The `Exit` exception will be retained as-is for now.
+
+## The Future
+
+The remaining bit of ambiguity is in whether or not `ConsoleHost#exit` should
+call `process.exit` directly, or throw the special `Exit` error. As discussed,
+the latter is somewhat valuable as a mechanism to allow graceful shutdowns.
+This motivates extending `Exit` to take an exit code, and making the host
+throw it. However, this removes some responsibility from the host object
```

Comment:
What if the runtime threw the `Exit` exception (with a status code) and the `Cli` class called `host.exit`? Then the host would still own exiting while also allowing for graceful cleanup.